## Import Python Library
1. You may need to install torch_geometric and the corresponding packages. Please run pip install requirements.txt first
2. If you are installing it for the first time, it will take a while
3. We are using a nvidia GPU for our experiment.

In [1]:
import torch
import os
import math
import numpy as np
import pandas as pd
import yaml
from utils.helper import print_loss_stat, seed_everything, eval_results

In [2]:
from loader import load_data, load_unsuper_data
from model import TLModel
from tqdm import tqdm

### prepare data
1. We have upload the dataset in the google drive: [here](https://drive.google.com/drive/folders/1ps7xXsIONxvUJu9nOvD_9_hl9W9aI3V7?usp=drive_link)
2. We didn't include the dataset into our github repo because of the size issue


### Define the yaml file
+ **Dataset**, Put the data into the datasets
+ **Model**:
    + Input_Dim: is related with the dataset
    + Conv: define the graph Convolution method. Specifically, please choose “TransformerConv” for running the SpaRx model, “GATConv” for the SpaRx-GAT, and “GCNConv” for the SpaRx-GCN model 
    + NUM_HIDDENS: the hidden dims for the first layer and the second layer
    + num_classes: 2 categories, one for the source, another for the target

In [3]:
yaml_file = 'configure_default.yml'
with open(yaml_file) as f:
    cfg = yaml.load(f, Loader=yaml.FullLoader)

seed_everything(cfg['SEED'])

### train the transfer model
Define the settings in the yaml file:

+ **lr:**: the learning rate
+ **Epochs:** training epochs
+ **Save_path:** the checkpoint path
+ **Momentum** and **Weight_dacay**: parameters for the Adam Optimizer
+ **grad_clip:** clip the gradient when it was too large



In [4]:
def train(cfg):
    # build data
    src_data_root = cfg['DATASET']['Source_data_root']
    src_label_root = cfg['DATASET']['Source_label_root']
    src_adj_root = cfg['DATASET']['Source_adj_root']

    src, src_train_idx, src_test_idx = load_data(root=src_data_root,
                                    adj_root=src_adj_root,
                                    label_root=src_label_root,
                                    num_val=0.2, random_state=cfg['SEED'])
    
    tar_data_root = cfg['DATASET']['target_data_root']
    tar_label_root = cfg['DATASET']['target_label_root']
    tar_adj_root = cfg['DATASET']['target_adj_root']
    tar = load_unsuper_data(root=tar_data_root,
                            adj_root=tar_adj_root,
                            label_root=tar_label_root)

    best_loss = 10000000
    best_f1 = 0
    best_epoch = 0
    # # build up model
    model = TLModel(cfg)
    for epoch in tqdm(range(cfg['TRAIN']['Epochs'])):
        # print('best_F1 is %f in Epoch %d' %(best_f1, best_epoch))
        model.set_input(src, tar)
        model.update_parameters()
        loss_stat = model.get_current_loss()
        loss = loss_stat['overall_loss'].item()

        # print_loss_stat(loss_stat, epoch, cfg['TRAIN']['Epochs'])

        f1 = val(cfg, model, tar)
        print(f1)
        if best_f1 < f1:
            model.save('best_f1')
            best_f1 = f1
            best_epoch = epoch
        if best_loss > loss:
            model.save('best_loss')
            best_loss = loss
            best_epoch = epoch
    model.save("last")
    #src_emb, tar_emb = model.get_emb()
    f1 = val(cfg, model, tar)
    print('f1 is %.4f'%(f1))

In [5]:
@torch.no_grad()
def val(cfg, model, tar):
    model.set_input(tar, tar)
    pred = model.inference()
    pred = pred.detach().cpu()
    label = tar.y.detach().cpu()

    f1 = eval_results(pred, label, cfg['TEST']['Verbose'])
    print("f1 is %.4f"%(f1))
    return f1

### training details
We train SpaRx for 100 epochs. For the demo, we print the performance for each epoch. At the very begining (first 30 epochs), the training model didn't learn the target infos. But the performance grows gradually and the final F1 score achieved to 0.9633.

In [6]:
train(cfg)

Using TransformerConv


  0%|                                                   | 0/100 [00:00<?, ?it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


  2%|▊                                          | 2/100 [00:01<01:23,  1.17it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


  3%|█▎                                         | 3/100 [00:02<00:56,  1.71it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)

------------------------results----------------------
         1	      1000
      1000	      1000
      acc:	    0.5005
precision:	    1.0000
   recall:	    0.0010
Specificity:	    1.0000
Sensitivity:	    0.0010
F1-measure:	    0.0020
------------------------------------------------------

f1 is 0.0020
tensor(0.0020)


  4%|█▋                                         | 4/100 [00:02<00:54,  1.75it/s]


------------------------results----------------------
         3	      1000
      1000	      1000
      acc:	    0.5015
precision:	    1.0000
   recall:	    0.0030
Specificity:	    1.0000
Sensitivity:	    0.0030
F1-measure:	    0.0060
------------------------------------------------------

f1 is 0.0060
tensor(0.0060)


  5%|██▏                                        | 5/100 [00:03<00:52,  1.80it/s]


------------------------results----------------------
         4	      1000
      1000	      1000
      acc:	    0.5020
precision:	    1.0000
   recall:	    0.0040
Specificity:	    1.0000
Sensitivity:	    0.0040
F1-measure:	    0.0080
------------------------------------------------------

f1 is 0.0080
tensor(0.0080)


  6%|██▌                                        | 6/100 [00:03<00:51,  1.81it/s]


------------------------results----------------------
         5	      1000
      1000	      1000
      acc:	    0.5025
precision:	    1.0000
   recall:	    0.0050
Specificity:	    1.0000
Sensitivity:	    0.0050
F1-measure:	    0.0100
------------------------------------------------------

f1 is 0.0100
tensor(0.0100)


  8%|███▍                                       | 8/100 [00:04<00:42,  2.17it/s]


------------------------results----------------------
         2	      1000
      1000	      1000
      acc:	    0.5010
precision:	    1.0000
   recall:	    0.0020
Specificity:	    1.0000
Sensitivity:	    0.0020
F1-measure:	    0.0040
------------------------------------------------------

f1 is 0.0040
tensor(0.0040)


  9%|███▊                                       | 9/100 [00:04<00:36,  2.48it/s]


------------------------results----------------------
         1	      1000
      1000	      1000
      acc:	    0.5005
precision:	    1.0000
   recall:	    0.0010
Specificity:	    1.0000
Sensitivity:	    0.0010
F1-measure:	    0.0020
------------------------------------------------------

f1 is 0.0020
tensor(0.0020)


 10%|████▏                                     | 10/100 [00:05<00:32,  2.77it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 11%|████▌                                     | 11/100 [00:05<00:29,  2.99it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 12%|█████                                     | 12/100 [00:05<00:27,  3.16it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 13%|█████▍                                    | 13/100 [00:06<00:26,  3.30it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)

------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 15%|██████▎                                   | 15/100 [00:06<00:29,  2.87it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 16%|██████▋                                   | 16/100 [00:07<00:27,  3.07it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 17%|███████▏                                  | 17/100 [00:07<00:25,  3.26it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 18%|███████▌                                  | 18/100 [00:07<00:24,  3.35it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)

------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 20%|████████▍                                 | 20/100 [00:08<00:28,  2.84it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 21%|████████▊                                 | 21/100 [00:08<00:25,  3.06it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 22%|█████████▏                                | 22/100 [00:09<00:24,  3.25it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 23%|█████████▋                                | 23/100 [00:09<00:22,  3.39it/s]


------------------------results----------------------
         0	      1000
      1000	      1000
      acc:	    0.5000
precision:	    0.0000
   recall:	    0.0000
Specificity:	    1.0000
Sensitivity:	    0.0000
F1-measure:	       nan
------------------------------------------------------

f1 is nan
tensor(nan)


 24%|██████████                                | 24/100 [00:09<00:21,  3.50it/s]


------------------------results----------------------
         1	      1000
      1000	      1000
      acc:	    0.5005
precision:	    1.0000
   recall:	    0.0010
Specificity:	    1.0000
Sensitivity:	    0.0010
F1-measure:	    0.0020
------------------------------------------------------

f1 is 0.0020
tensor(0.0020)

------------------------results----------------------
        13	      1000
      1000	      1000
      acc:	    0.5065
precision:	    1.0000
   recall:	    0.0130
Specificity:	    1.0000
Sensitivity:	    0.0130
F1-measure:	    0.0257
------------------------------------------------------

f1 is 0.0257
tensor(0.0257)


 25%|██████████▌                               | 25/100 [00:10<00:27,  2.77it/s]


------------------------results----------------------
        49	      1000
      1000	      1000
      acc:	    0.5245
precision:	    1.0000
   recall:	    0.0490
Specificity:	    1.0000
Sensitivity:	    0.0490
F1-measure:	    0.0934
------------------------------------------------------

f1 is 0.0934
tensor(0.0934)


 26%|██████████▉                               | 26/100 [00:10<00:30,  2.45it/s]


------------------------results----------------------
        66	      1000
      1000	      1000
      acc:	    0.5330
precision:	    1.0000
   recall:	    0.0660
Specificity:	    1.0000
Sensitivity:	    0.0660
F1-measure:	    0.1238
------------------------------------------------------

f1 is 0.1238
tensor(0.1238)


 27%|███████████▎                              | 27/100 [00:11<00:32,  2.23it/s]


------------------------results----------------------
        72	      1000
      1000	      1000
      acc:	    0.5360
precision:	    1.0000
   recall:	    0.0720
Specificity:	    1.0000
Sensitivity:	    0.0720
F1-measure:	    0.1343
------------------------------------------------------

f1 is 0.1343
tensor(0.1343)


 29%|████████████▏                             | 29/100 [00:12<00:33,  2.15it/s]


------------------------results----------------------
        62	      1000
      1000	      1000
      acc:	    0.5310
precision:	    1.0000
   recall:	    0.0620
Specificity:	    1.0000
Sensitivity:	    0.0620
F1-measure:	    0.1168
------------------------------------------------------

f1 is 0.1168
tensor(0.1168)


 30%|████████████▌                             | 30/100 [00:12<00:28,  2.46it/s]


------------------------results----------------------
        32	      1000
      1000	      1000
      acc:	    0.5160
precision:	    1.0000
   recall:	    0.0320
Specificity:	    1.0000
Sensitivity:	    0.0320
F1-measure:	    0.0620
------------------------------------------------------

f1 is 0.0620
tensor(0.0620)


 31%|█████████████                             | 31/100 [00:12<00:25,  2.75it/s]


------------------------results----------------------
         7	      1000
      1000	      1000
      acc:	    0.5035
precision:	    1.0000
   recall:	    0.0070
Specificity:	    1.0000
Sensitivity:	    0.0070
F1-measure:	    0.0139
------------------------------------------------------

f1 is 0.0139
tensor(0.0139)


 32%|█████████████▍                            | 32/100 [00:13<00:22,  2.96it/s]


------------------------results----------------------
         2	      1000
      1000	      1000
      acc:	    0.5010
precision:	    1.0000
   recall:	    0.0020
Specificity:	    1.0000
Sensitivity:	    0.0020
F1-measure:	    0.0040
------------------------------------------------------

f1 is 0.0040
tensor(0.0040)


 33%|█████████████▊                            | 33/100 [00:13<00:21,  3.15it/s]


------------------------results----------------------
         1	      1000
      1000	      1000
      acc:	    0.5005
precision:	    1.0000
   recall:	    0.0010
Specificity:	    1.0000
Sensitivity:	    0.0010
F1-measure:	    0.0020
------------------------------------------------------

f1 is 0.0020
tensor(0.0020)


 34%|██████████████▎                           | 34/100 [00:13<00:19,  3.31it/s]


------------------------results----------------------
         9	      1000
      1000	      1000
      acc:	    0.5045
precision:	    1.0000
   recall:	    0.0090
Specificity:	    1.0000
Sensitivity:	    0.0090
F1-measure:	    0.0178
------------------------------------------------------

f1 is 0.0178
tensor(0.0178)


 35%|██████████████▋                           | 35/100 [00:13<00:18,  3.45it/s]


------------------------results----------------------
        25	      1000
       999	      1000
      acc:	    0.5120
precision:	    0.9615
   recall:	    0.0250
Specificity:	    0.9990
Sensitivity:	    0.0250
F1-measure:	    0.0487
------------------------------------------------------

f1 is 0.0487
tensor(0.0487)


 36%|███████████████                           | 36/100 [00:14<00:18,  3.53it/s]


------------------------results----------------------
        48	      1000
       999	      1000
      acc:	    0.5235
precision:	    0.9796
   recall:	    0.0480
Specificity:	    0.9990
Sensitivity:	    0.0480
F1-measure:	    0.0915
------------------------------------------------------

f1 is 0.0915
tensor(0.0915)

------------------------results----------------------
        78	      1000
      1000	      1000
      acc:	    0.5390
precision:	    1.0000
   recall:	    0.0780
Specificity:	    1.0000
Sensitivity:	    0.0780
F1-measure:	    0.1447
------------------------------------------------------

f1 is 0.1447
tensor(0.1447)


 37%|███████████████▌                          | 37/100 [00:14<00:23,  2.63it/s]


------------------------results----------------------
       102	      1000
      1000	      1000
      acc:	    0.5510
precision:	    1.0000
   recall:	    0.1020
Specificity:	    1.0000
Sensitivity:	    0.1020
F1-measure:	    0.1851
------------------------------------------------------

f1 is 0.1851
tensor(0.1851)


 38%|███████████████▉                          | 38/100 [00:15<00:27,  2.24it/s]


------------------------results----------------------
       113	      1000
      1000	      1000
      acc:	    0.5565
precision:	    1.0000
   recall:	    0.1130
Specificity:	    1.0000
Sensitivity:	    0.1130
F1-measure:	    0.2031
------------------------------------------------------

f1 is 0.2031
tensor(0.2031)


 40%|████████████████▊                         | 40/100 [00:16<00:25,  2.38it/s]


------------------------results----------------------
        97	      1000
       999	      1000
      acc:	    0.5480
precision:	    0.9898
   recall:	    0.0970
Specificity:	    0.9990
Sensitivity:	    0.0970
F1-measure:	    0.1767
------------------------------------------------------

f1 is 0.1767
tensor(0.1767)


 41%|█████████████████▏                        | 41/100 [00:16<00:22,  2.66it/s]


------------------------results----------------------
        91	      1000
       999	      1000
      acc:	    0.5450
precision:	    0.9891
   recall:	    0.0910
Specificity:	    0.9990
Sensitivity:	    0.0910
F1-measure:	    0.1667
------------------------------------------------------

f1 is 0.1667
tensor(0.1667)


 42%|█████████████████▋                        | 42/100 [00:16<00:19,  2.91it/s]


------------------------results----------------------
        87	      1000
       997	      1000
      acc:	    0.5420
precision:	    0.9667
   recall:	    0.0870
Specificity:	    0.9970
Sensitivity:	    0.0870
F1-measure:	    0.1596
------------------------------------------------------

f1 is 0.1596
tensor(0.1596)


 43%|██████████████████                        | 43/100 [00:16<00:18,  3.08it/s]


------------------------results----------------------
        83	      1000
       998	      1000
      acc:	    0.5405
precision:	    0.9765
   recall:	    0.0830
Specificity:	    0.9980
Sensitivity:	    0.0830
F1-measure:	    0.1530
------------------------------------------------------

f1 is 0.1530
tensor(0.1530)


 44%|██████████████████▍                       | 44/100 [00:17<00:17,  3.24it/s]


------------------------results----------------------
       100	      1000
       999	      1000
      acc:	    0.5495
precision:	    0.9901
   recall:	    0.1000
Specificity:	    0.9990
Sensitivity:	    0.1000
F1-measure:	    0.1817
------------------------------------------------------

f1 is 0.1817
tensor(0.1817)

------------------------results----------------------
       150	      1000
       999	      1000
      acc:	    0.5745
precision:	    0.9934
   recall:	    0.1500
Specificity:	    0.9990
Sensitivity:	    0.1500
F1-measure:	    0.2606
------------------------------------------------------

f1 is 0.2606
tensor(0.2606)


 45%|██████████████████▉                       | 45/100 [00:17<00:20,  2.65it/s]


------------------------results----------------------
       196	      1000
       999	      1000
      acc:	    0.5975
precision:	    0.9949
   recall:	    0.1960
Specificity:	    0.9990
Sensitivity:	    0.1960
F1-measure:	    0.3275
------------------------------------------------------

f1 is 0.3275
tensor(0.3275)


 46%|███████████████████▎                      | 46/100 [00:18<00:22,  2.38it/s]


------------------------results----------------------
       246	      1000
       999	      1000
      acc:	    0.6225
precision:	    0.9960
   recall:	    0.2460
Specificity:	    0.9990
Sensitivity:	    0.2460
F1-measure:	    0.3945
------------------------------------------------------

f1 is 0.3945
tensor(0.3945)


 47%|███████████████████▋                      | 47/100 [00:18<00:23,  2.23it/s]


------------------------results----------------------
       268	      1000
       999	      1000
      acc:	    0.6335
precision:	    0.9963
   recall:	    0.2680
Specificity:	    0.9990
Sensitivity:	    0.2680
F1-measure:	    0.4224
------------------------------------------------------

f1 is 0.4224
tensor(0.4224)


 49%|████████████████████▌                     | 49/100 [00:19<00:23,  2.15it/s]


------------------------results----------------------
       235	      1000
       999	      1000
      acc:	    0.6170
precision:	    0.9958
   recall:	    0.2350
Specificity:	    0.9990
Sensitivity:	    0.2350
F1-measure:	    0.3803
------------------------------------------------------

f1 is 0.3803
tensor(0.3803)


 50%|█████████████████████                     | 50/100 [00:20<00:20,  2.46it/s]


------------------------results----------------------
       202	      1000
       998	      1000
      acc:	    0.6000
precision:	    0.9902
   recall:	    0.2020
Specificity:	    0.9980
Sensitivity:	    0.2020
F1-measure:	    0.3355
------------------------------------------------------

f1 is 0.3355
tensor(0.3355)


 51%|█████████████████████▍                    | 51/100 [00:20<00:17,  2.73it/s]


------------------------results----------------------
       165	      1000
       998	      1000
      acc:	    0.5815
precision:	    0.9880
   recall:	    0.1650
Specificity:	    0.9980
Sensitivity:	    0.1650
F1-measure:	    0.2828
------------------------------------------------------

f1 is 0.2828
tensor(0.2828)


 52%|█████████████████████▊                    | 52/100 [00:20<00:16,  2.96it/s]


------------------------results----------------------
       183	      1000
       998	      1000
      acc:	    0.5905
precision:	    0.9892
   recall:	    0.1830
Specificity:	    0.9980
Sensitivity:	    0.1830
F1-measure:	    0.3089
------------------------------------------------------

f1 is 0.3089
tensor(0.3089)


 53%|██████████████████████▎                   | 53/100 [00:20<00:14,  3.16it/s]


------------------------results----------------------
       158	      1000
      1000	      1000
      acc:	    0.5790
precision:	    1.0000
   recall:	    0.1580
Specificity:	    1.0000
Sensitivity:	    0.1580
F1-measure:	    0.2729
------------------------------------------------------

f1 is 0.2729
tensor(0.2729)


 54%|██████████████████████▋                   | 54/100 [00:21<00:13,  3.33it/s]


------------------------results----------------------
       122	      1000
      1000	      1000
      acc:	    0.5610
precision:	    1.0000
   recall:	    0.1220
Specificity:	    1.0000
Sensitivity:	    0.1220
F1-measure:	    0.2175
------------------------------------------------------

f1 is 0.2175
tensor(0.2175)


 55%|███████████████████████                   | 55/100 [00:21<00:13,  3.43it/s]


------------------------results----------------------
       110	      1000
      1000	      1000
      acc:	    0.5550
precision:	    1.0000
   recall:	    0.1100
Specificity:	    1.0000
Sensitivity:	    0.1100
F1-measure:	    0.1982
------------------------------------------------------

f1 is 0.1982
tensor(0.1982)


 56%|███████████████████████▌                  | 56/100 [00:21<00:12,  3.51it/s]


------------------------results----------------------
        96	      1000
      1000	      1000
      acc:	    0.5480
precision:	    1.0000
   recall:	    0.0960
Specificity:	    1.0000
Sensitivity:	    0.0960
F1-measure:	    0.1752
------------------------------------------------------

f1 is 0.1752
tensor(0.1752)


 57%|███████████████████████▉                  | 57/100 [00:22<00:12,  3.58it/s]


------------------------results----------------------
        90	      1000
      1000	      1000
      acc:	    0.5450
precision:	    1.0000
   recall:	    0.0900
Specificity:	    1.0000
Sensitivity:	    0.0900
F1-measure:	    0.1651
------------------------------------------------------

f1 is 0.1651
tensor(0.1651)


 58%|████████████████████████▎                 | 58/100 [00:22<00:11,  3.62it/s]


------------------------results----------------------
       108	      1000
      1000	      1000
      acc:	    0.5540
precision:	    1.0000
   recall:	    0.1080
Specificity:	    1.0000
Sensitivity:	    0.1080
F1-measure:	    0.1949
------------------------------------------------------

f1 is 0.1949
tensor(0.1949)


 59%|████████████████████████▊                 | 59/100 [00:22<00:11,  3.65it/s]


------------------------results----------------------
       124	      1000
      1000	      1000
      acc:	    0.5620
precision:	    1.0000
   recall:	    0.1240
Specificity:	    1.0000
Sensitivity:	    0.1240
F1-measure:	    0.2206
------------------------------------------------------

f1 is 0.2206
tensor(0.2206)


 60%|█████████████████████████▏                | 60/100 [00:22<00:10,  3.69it/s]


------------------------results----------------------
       187	      1000
       999	      1000
      acc:	    0.5930
precision:	    0.9947
   recall:	    0.1870
Specificity:	    0.9990
Sensitivity:	    0.1870
F1-measure:	    0.3148
------------------------------------------------------

f1 is 0.3148
tensor(0.3148)

------------------------results----------------------
       275	      1000
       999	      1000
      acc:	    0.6370
precision:	    0.9964
   recall:	    0.2750
Specificity:	    0.9990
Sensitivity:	    0.2750
F1-measure:	    0.4310
------------------------------------------------------

f1 is 0.4310
tensor(0.4310)


 61%|█████████████████████████▌                | 61/100 [00:23<00:14,  2.71it/s]


------------------------results----------------------
       330	      1000
       998	      1000
      acc:	    0.6640
precision:	    0.9940
   recall:	    0.3300
Specificity:	    0.9980
Sensitivity:	    0.3300
F1-measure:	    0.4955
------------------------------------------------------

f1 is 0.4955
tensor(0.4955)


 62%|██████████████████████████                | 62/100 [00:24<00:18,  2.02it/s]


------------------------results----------------------
       364	      1000
       998	      1000
      acc:	    0.6810
precision:	    0.9945
   recall:	    0.3640
Specificity:	    0.9980
Sensitivity:	    0.3640
F1-measure:	    0.5329
------------------------------------------------------

f1 is 0.5329
tensor(0.5329)


 63%|██████████████████████████▍               | 63/100 [00:25<00:21,  1.71it/s]


------------------------results----------------------
       375	      1000
       998	      1000
      acc:	    0.6865
precision:	    0.9947
   recall:	    0.3750
Specificity:	    0.9980
Sensitivity:	    0.3750
F1-measure:	    0.5447
------------------------------------------------------

f1 is 0.5447
tensor(0.5447)


 65%|███████████████████████████▎              | 65/100 [00:25<00:16,  2.09it/s]


------------------------results----------------------
       374	      1000
       998	      1000
      acc:	    0.6860
precision:	    0.9947
   recall:	    0.3740
Specificity:	    0.9980
Sensitivity:	    0.3740
F1-measure:	    0.5436
------------------------------------------------------

f1 is 0.5436
tensor(0.5436)


 66%|███████████████████████████▋              | 66/100 [00:26<00:14,  2.41it/s]


------------------------results----------------------
       371	      1000
       998	      1000
      acc:	    0.6845
precision:	    0.9946
   recall:	    0.3710
Specificity:	    0.9980
Sensitivity:	    0.3710
F1-measure:	    0.5404
------------------------------------------------------

f1 is 0.5404
tensor(0.5404)

------------------------results----------------------
       432	      1000
       998	      1000
      acc:	    0.7150
precision:	    0.9954
   recall:	    0.4320
Specificity:	    0.9980
Sensitivity:	    0.4320
F1-measure:	    0.6025
------------------------------------------------------

f1 is 0.6025
tensor(0.6025)


 67%|████████████████████████████▏             | 67/100 [00:26<00:14,  2.23it/s]


------------------------results----------------------
       528	      1000
       998	      1000
      acc:	    0.7630
precision:	    0.9962
   recall:	    0.5280
Specificity:	    0.9980
Sensitivity:	    0.5280
F1-measure:	    0.6902
------------------------------------------------------

f1 is 0.6902
tensor(0.6902)


 68%|████████████████████████████▌             | 68/100 [00:27<00:15,  2.10it/s]


------------------------results----------------------
       672	      1000
       997	      1000
      acc:	    0.8345
precision:	    0.9956
   recall:	    0.6720
Specificity:	    0.9970
Sensitivity:	    0.6720
F1-measure:	    0.8024
------------------------------------------------------

f1 is 0.8024
tensor(0.8024)


 69%|████████████████████████████▉             | 69/100 [00:27<00:15,  2.01it/s]


------------------------results----------------------
       781	      1000
       997	      1000
      acc:	    0.8890
precision:	    0.9962
   recall:	    0.7810
Specificity:	    0.9970
Sensitivity:	    0.7810
F1-measure:	    0.8756
------------------------------------------------------

f1 is 0.8756
tensor(0.8756)


 70%|█████████████████████████████▍            | 70/100 [00:28<00:15,  1.93it/s]


------------------------results----------------------
       842	      1000
       992	      1000
      acc:	    0.9170
precision:	    0.9906
   recall:	    0.8420
Specificity:	    0.9920
Sensitivity:	    0.8420
F1-measure:	    0.9103
------------------------------------------------------

f1 is 0.9103
tensor(0.9103)


 71%|█████████████████████████████▊            | 71/100 [00:28<00:15,  1.90it/s]


------------------------results----------------------
       897	      1000
       986	      1000
      acc:	    0.9415
precision:	    0.9846
   recall:	    0.8970
Specificity:	    0.9860
Sensitivity:	    0.8970
F1-measure:	    0.9388
------------------------------------------------------

f1 is 0.9388
tensor(0.9388)


 72%|██████████████████████████████▏           | 72/100 [00:29<00:16,  1.66it/s]


------------------------results----------------------
       924	      1000
       974	      1000
      acc:	    0.9490
precision:	    0.9726
   recall:	    0.9240
Specificity:	    0.9740
Sensitivity:	    0.9240
F1-measure:	    0.9477
------------------------------------------------------

f1 is 0.9477
tensor(0.9477)


 73%|██████████████████████████████▋           | 73/100 [00:30<00:15,  1.72it/s]


------------------------results----------------------
       916	      1000
       978	      1000
      acc:	    0.9470
precision:	    0.9765
   recall:	    0.9160
Specificity:	    0.9780
Sensitivity:	    0.9160
F1-measure:	    0.9453
------------------------------------------------------

f1 is 0.9453
tensor(0.9453)


 75%|███████████████████████████████▌          | 75/100 [00:30<00:11,  2.09it/s]


------------------------results----------------------
       856	      1000
       994	      1000
      acc:	    0.9250
precision:	    0.9930
   recall:	    0.8560
Specificity:	    0.9940
Sensitivity:	    0.8560
F1-measure:	    0.9194
------------------------------------------------------

f1 is 0.9194
tensor(0.9194)


 76%|███████████████████████████████▉          | 76/100 [00:31<00:09,  2.42it/s]


------------------------results----------------------
       842	      1000
       994	      1000
      acc:	    0.9180
precision:	    0.9929
   recall:	    0.8420
Specificity:	    0.9940
Sensitivity:	    0.8420
F1-measure:	    0.9113
------------------------------------------------------

f1 is 0.9113
tensor(0.9113)


 77%|████████████████████████████████▎         | 77/100 [00:31<00:08,  2.71it/s]


------------------------results----------------------
       847	      1000
       994	      1000
      acc:	    0.9205
precision:	    0.9930
   recall:	    0.8470
Specificity:	    0.9940
Sensitivity:	    0.8470
F1-measure:	    0.9142
------------------------------------------------------

f1 is 0.9142
tensor(0.9142)


 78%|████████████████████████████████▊         | 78/100 [00:31<00:07,  2.97it/s]


------------------------results----------------------
       871	      1000
       994	      1000
      acc:	    0.9325
precision:	    0.9932
   recall:	    0.8710
Specificity:	    0.9940
Sensitivity:	    0.8710
F1-measure:	    0.9281
------------------------------------------------------

f1 is 0.9281
tensor(0.9281)


 79%|█████████████████████████████████▏        | 79/100 [00:31<00:06,  3.17it/s]


------------------------results----------------------
       875	      1000
       992	      1000
      acc:	    0.9335
precision:	    0.9909
   recall:	    0.8750
Specificity:	    0.9920
Sensitivity:	    0.8750
F1-measure:	    0.9294
------------------------------------------------------

f1 is 0.9294
tensor(0.9294)


 80%|█████████████████████████████████▌        | 80/100 [00:32<00:06,  3.32it/s]


------------------------results----------------------
       882	      1000
       989	      1000
      acc:	    0.9355
precision:	    0.9877
   recall:	    0.8820
Specificity:	    0.9890
Sensitivity:	    0.8820
F1-measure:	    0.9319
------------------------------------------------------

f1 is 0.9319
tensor(0.9319)


 81%|██████████████████████████████████        | 81/100 [00:32<00:05,  3.46it/s]


------------------------results----------------------
       847	      1000
       993	      1000
      acc:	    0.9200
precision:	    0.9918
   recall:	    0.8470
Specificity:	    0.9930
Sensitivity:	    0.8470
F1-measure:	    0.9137
------------------------------------------------------

f1 is 0.9137
tensor(0.9137)


 82%|██████████████████████████████████▍       | 82/100 [00:32<00:05,  3.53it/s]


------------------------results----------------------
       821	      1000
       994	      1000
      acc:	    0.9075
precision:	    0.9927
   recall:	    0.8210
Specificity:	    0.9940
Sensitivity:	    0.8210
F1-measure:	    0.8987
------------------------------------------------------

f1 is 0.8987
tensor(0.8987)


 83%|██████████████████████████████████▊       | 83/100 [00:33<00:04,  3.58it/s]


------------------------results----------------------
       760	      1000
       996	      1000
      acc:	    0.8780
precision:	    0.9948
   recall:	    0.7600
Specificity:	    0.9960
Sensitivity:	    0.7600
F1-measure:	    0.8617
------------------------------------------------------

f1 is 0.8617
tensor(0.8617)


 84%|███████████████████████████████████▎      | 84/100 [00:33<00:04,  3.61it/s]


------------------------results----------------------
       719	      1000
       997	      1000
      acc:	    0.8580
precision:	    0.9958
   recall:	    0.7190
Specificity:	    0.9970
Sensitivity:	    0.7190
F1-measure:	    0.8351
------------------------------------------------------

f1 is 0.8351
tensor(0.8351)


 85%|███████████████████████████████████▋      | 85/100 [00:33<00:04,  3.61it/s]


------------------------results----------------------
       682	      1000
       997	      1000
      acc:	    0.8395
precision:	    0.9956
   recall:	    0.6820
Specificity:	    0.9970
Sensitivity:	    0.6820
F1-measure:	    0.8095
------------------------------------------------------

f1 is 0.8095
tensor(0.8095)


 86%|████████████████████████████████████      | 86/100 [00:33<00:03,  3.66it/s]


------------------------results----------------------
       683	      1000
       997	      1000
      acc:	    0.8400
precision:	    0.9956
   recall:	    0.6830
Specificity:	    0.9970
Sensitivity:	    0.6830
F1-measure:	    0.8102
------------------------------------------------------

f1 is 0.8102
tensor(0.8102)


 87%|████████████████████████████████████▌     | 87/100 [00:34<00:03,  3.70it/s]


------------------------results----------------------
       665	      1000
       997	      1000
      acc:	    0.8310
precision:	    0.9955
   recall:	    0.6650
Specificity:	    0.9970
Sensitivity:	    0.6650
F1-measure:	    0.7974
------------------------------------------------------

f1 is 0.7974
tensor(0.7974)


 88%|████████████████████████████████████▉     | 88/100 [00:34<00:03,  3.72it/s]


------------------------results----------------------
       650	      1000
       996	      1000
      acc:	    0.8230
precision:	    0.9939
   recall:	    0.6500
Specificity:	    0.9960
Sensitivity:	    0.6500
F1-measure:	    0.7860
------------------------------------------------------

f1 is 0.7860
tensor(0.7860)


 89%|█████████████████████████████████████▍    | 89/100 [00:34<00:02,  3.70it/s]


------------------------results----------------------
       677	      1000
       994	      1000
      acc:	    0.8355
precision:	    0.9912
   recall:	    0.6770
Specificity:	    0.9940
Sensitivity:	    0.6770
F1-measure:	    0.8045
------------------------------------------------------

f1 is 0.8045
tensor(0.8045)


 90%|█████████████████████████████████████▊    | 90/100 [00:34<00:02,  3.69it/s]


------------------------results----------------------
       736	      1000
       994	      1000
      acc:	    0.8650
precision:	    0.9919
   recall:	    0.7360
Specificity:	    0.9940
Sensitivity:	    0.7360
F1-measure:	    0.8450
------------------------------------------------------

f1 is 0.8450
tensor(0.8450)

------------------------results----------------------
       776	      1000
       994	      1000
      acc:	    0.8850
precision:	    0.9923
   recall:	    0.7760
Specificity:	    0.9940
Sensitivity:	    0.7760
F1-measure:	    0.8709
------------------------------------------------------

f1 is 0.8709
tensor(0.8709)


 91%|██████████████████████████████████████▏   | 91/100 [00:35<00:03,  2.83it/s]


------------------------results----------------------
       809	      1000
       996	      1000
      acc:	    0.9025
precision:	    0.9951
   recall:	    0.8090
Specificity:	    0.9960
Sensitivity:	    0.8090
F1-measure:	    0.8924
------------------------------------------------------

f1 is 0.8924
tensor(0.8924)


 93%|███████████████████████████████████████   | 93/100 [00:36<00:02,  2.75it/s]


------------------------results----------------------
       841	      1000
       994	      1000
      acc:	    0.9175
precision:	    0.9929
   recall:	    0.8410
Specificity:	    0.9940
Sensitivity:	    0.8410
F1-measure:	    0.9107
------------------------------------------------------

f1 is 0.9107
tensor(0.9107)


 94%|███████████████████████████████████████▍  | 94/100 [00:36<00:02,  2.98it/s]


------------------------results----------------------
       865	      1000
       994	      1000
      acc:	    0.9295
precision:	    0.9931
   recall:	    0.8650
Specificity:	    0.9940
Sensitivity:	    0.8650
F1-measure:	    0.9246
------------------------------------------------------

f1 is 0.9246
tensor(0.9246)


 95%|███████████████████████████████████████▉  | 95/100 [00:36<00:01,  3.16it/s]


------------------------results----------------------
       880	      1000
       993	      1000
      acc:	    0.9365
precision:	    0.9921
   recall:	    0.8800
Specificity:	    0.9930
Sensitivity:	    0.8800
F1-measure:	    0.9327
------------------------------------------------------

f1 is 0.9327
tensor(0.9327)


 96%|████████████████████████████████████████▎ | 96/100 [00:37<00:01,  3.31it/s]


------------------------results----------------------
       894	      1000
       990	      1000
      acc:	    0.9420
precision:	    0.9889
   recall:	    0.8940
Specificity:	    0.9900
Sensitivity:	    0.8940
F1-measure:	    0.9391
------------------------------------------------------

f1 is 0.9391
tensor(0.9391)

------------------------results----------------------
       929	      1000
       986	      1000
      acc:	    0.9575
precision:	    0.9852
   recall:	    0.9290
Specificity:	    0.9860
Sensitivity:	    0.9290
F1-measure:	    0.9563
------------------------------------------------------

f1 is 0.9563
tensor(0.9563)


 97%|████████████████████████████████████████▋ | 97/100 [00:37<00:01,  2.66it/s]


------------------------results----------------------
       958	      1000
       969	      1000
      acc:	    0.9635
precision:	    0.9687
   recall:	    0.9580
Specificity:	    0.9690
Sensitivity:	    0.9580
F1-measure:	    0.9633
------------------------------------------------------

f1 is 0.9633
tensor(0.9633)


 99%|█████████████████████████████████████████▌| 99/100 [00:38<00:00,  2.62it/s]


------------------------results----------------------
       965	      1000
       954	      1000
      acc:	    0.9595
precision:	    0.9545
   recall:	    0.9650
Specificity:	    0.9540
Sensitivity:	    0.9650
F1-measure:	    0.9597
------------------------------------------------------

f1 is 0.9597
tensor(0.9597)


100%|█████████████████████████████████████████| 100/100 [00:38<00:00,  2.58it/s]


------------------------results----------------------
       969	      1000
       956	      1000
      acc:	    0.9625
precision:	    0.9566
   recall:	    0.9690
Specificity:	    0.9560
Sensitivity:	    0.9690
F1-measure:	    0.9627
------------------------------------------------------

f1 is 0.9627
tensor(0.9627)



------------------------results----------------------
       969	      1000
       956	      1000
      acc:	    0.9625
precision:	    0.9566
   recall:	    0.9690
Specificity:	    0.9560
Sensitivity:	    0.9690
F1-measure:	    0.9627
------------------------------------------------------

f1 is 0.9627
f1 is 0.9627


### Test the Model performance
In this demo, We first test the model with the best f1 score, which was the 969 epoch. After that, we save the embedding for downstream analysis

In [7]:
@torch.no_grad()
def test(cfg):
    tar_data_root = cfg['DATASET']['target_data_root']
    tar_label_root = cfg['DATASET']['target_label_root']
    tar_adj_root = cfg['DATASET']['target_adj_root']
    tar, tar_index = load_unsuper_data(root=tar_data_root,
                            adj_root=tar_adj_root,
                            label_root=tar_label_root, return_index=True)

    model = TLModel(cfg)
    model.load(cfg['TEST']['Save_path'])

    pred = model.set_input(tar, tar)
    pred = model.inference()
    # src_emb, tar_emb = model.get_emb()
    pred = pred.detach().cpu()

    label = tar.y.detach().cpu()
    f1 = eval_results(pred, label)

    # save pred to csv
    True_label = label.numpy()

    df = pd.DataFrame(index=tar_index)
    df['pred'] = pred.numpy()
    df['true_label'] = True_label

    # if not os.path.exists(os.path.join('results', cfg['Name'])):
    #     os.makedirs(os.path.join('results', cfg['Name']))
    
    if not os.path.join(cfg["TEST"]["Pred"]):
        os.makedirs(cfg["TEST"]["Pred"], exist_ok=True)
    
    df.to_csv(os.path.join(cfg['TEST']['Pred'],'prediction.csv'))        
    return f1

In [8]:
test(cfg)

Using TransformerConv


tensor(0.9633)

In [9]:
@torch.no_grad()
def save_emb(cfg):
    src_data_root = cfg['DATASET']['Source_data_root']
    src_label_root = cfg['DATASET']['Source_label_root']
    src_adj_root = cfg['DATASET']['Source_adj_root']

    src, src_index = load_unsuper_data(root=src_data_root,
                            adj_root=src_adj_root,
                            label_root=src_label_root, return_index=True)

    tar_data_root = cfg['DATASET']['target_data_root']
    tar_label_root = cfg['DATASET']['target_label_root']
    tar_adj_root = cfg['DATASET']['target_adj_root']
    tar, tar_index = load_unsuper_data(root=tar_data_root,
                            adj_root=tar_adj_root,
                            label_root=tar_label_root, return_index=True)


    model = TLModel(cfg)
    model.load(cfg['TEST']['Save_path'])

    pred = model.set_input(src, tar)
    src_emb, tar_emb = model.get_emb()

    src_emb = src_emb.detach().cpu().numpy()
    tar_emb = tar_emb.detach().cpu().numpy()

    df = pd.DataFrame(src_emb, index=src_index)
    # if not os.path.exists(os.path.join('results', cfg['Name'])):
    #     os.makedirs(os.path.join('results', cfg['Name']))
    df.to_csv(os.path.join(cfg['TEST']['Pred'],'src_emb.csv'))        

    df = pd.DataFrame(tar_emb, index=tar_index)
    df.to_csv(os.path.join(cfg['TEST']['Pred'],'tar_emb.csv'))        

In [10]:
save_emb(cfg)

Using TransformerConv


### How to choose the model when we don't know the ground truth?
1. In real implementation, we didn't have the ground truth, and therefore we provided another model called best_loss.pth, which had the smallest training loss. This can have overfitting issues and the performance is 0.8924

In [11]:
cfg['TEST']['Save_path'] = "./checkpoint/best_loss.pth"
seed_everything(cfg['SEED'])

In [12]:
print(cfg['TEST']['Save_path'])

./checkpoint/best_loss.pth


In [13]:
test(cfg)

Using TransformerConv


tensor(0.8924)

2. We also present the result from the last epoch after training 100 times. The F1 score in the last layer is 0.9627, which is compariable with the best F1 model (0.9633)

In [14]:
cfg['TEST']['Save_path'] = "./checkpoint/last.pth"
seed_everything(cfg['SEED'])
print(cfg['TEST']['Save_path'])
test(cfg)

./checkpoint/last.pth
Using TransformerConv


tensor(0.9627)